src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1350,19 +1501,59 @@ DatabaseShardImp::getCache(std::uint32_t seq)
     return std::make_pair(pCache, nCache);
 }
 
-std::uint64_t
-DatabaseShardImp::available() const
+bool
+DatabaseShardImp::sufficientStorage(
+    std::uint32_t numShards,
+    PathDesignation pathDesignation,
+    std::lock_guard<std::mutex> const&) const
 {
     try
     {
-        return boost::filesystem::space(dir_).available;
+        std::vector<std::uint64_t> capacities;
+
+        if (pathDesignation == PathDesignation::historical &&
+            !historicalPaths_.empty())
+        {
+            capacities.reserve(historicalPaths_.size());
+
+            // Get the available storage for each historical path
+            std::for_each(
+                historicalPaths_.begin(),
+                historicalPaths_.end(),
+                [&capacities](boost::filesystem::path const& path) {
+                    capacities.push_back(
+                        boost::filesystem::space(path).available);
+                });
+        }
+        else
+        {
+            // Get the available storage for the main shard path
+            capacities.push_back(boost::filesystem::space(dir_).available);
+        }
+
+        for (std::uint64_t const capacity : capacities)
+        {
+            // Leverage all the historical shard paths to
+            // see if collectively they can fit the specified
+            // number of shards. For this to work properly,
+            // each historical path must correspond to a separate
+            // physical device or filesystem.
```

Comment:
It's not a perfect check, but how hard would it be to check for duplicates when you're populating `capacities`? If two different paths have the same number of bytes available, there's a chance that they're the same device. It's not guaranteed, so don't fail or anything, but you could log a warning message. Something like `JLOG(j_.warn()) << path1 << " and " << path2 << " both have " << capacity << " bytes free, and may be the same device. Each configured historical storage path should be on a unique device.`

You could even make it `#if ! BOOST_OS_LINUX`

Reply:
Done